### PR TITLE
API-29607-handle-null-values

### PR DESCRIPTION
**Pull request by stiehlrod**
## Summary

- Makes several attributes in the schema nullable, and makes the changes to validation and the mapper that are a result of the nullability.
- doc updates will be addressed in another PR to save lines of code per PR
- Additional tests will be added in another PR to save lines of code per PR
- json file updates

## Related issue(s)
- [API-29607](https://jira.devops.va.gov/browse/API-29607)

## Testing done

- rspec
- Postman {{uri}}/services/claims/v2/veterans/1008711084V002120/526
   request.body with minimal values:
```{
    "data": {
        "type": "form/526",
        "attributes": {
            "claimantCertification": true,
            "claimProcessType": "STANDARD_CLAIM_PROCESS",
            "veteranIdentification": {
                "currentVaEmployee": false,
                "mailingAddress": {
                    "addressLine1": "1234 Couch Street",
                    "addressLine2": "Unit 4",
                    "addressLine3": "Room 1",
                    "city": "Portland",
                    "country": "USA",
                    "zipFirstFive": "41726",
                    "zipLastFour": "1234",
                    "state": "OR"
                },
                "emailAddress": {
                    "email": "valid@somedomain.com"
                }
            },
            "changeOfAddress": {
                "dates": {
                    "beginDate": "11-29-2013",
                    "endDate": "10-11-2025"
                },
                "typeOfAddressChange": "TEMPORARY",
                "addressLine1": "10 Peach St",
                "addressLine2": "Unit 4",
                "addressLine3": "Room 1",
                "city": "Atlanta",
                "zipFirstFive": "42220",
                "zipLastFour": "9897",
                "state": "GA",
                "country": "USA"
            },
            "homeless": {
                "pointOfContact": "john stewart",
                "currentlyHomeless": {
                    "homelessSituationOptions": "FLEEING_CURRENT_RESIDENCE",
                    "otherDescription": "ABCDEFGHIJKLM"
                }
            },
            "disabilities": [
                {
                    "disabilityActionType": "NEW",
                    "name": "Hearing Loss",
                    "serviceRelevance": "Heavy equipment operator in service"
                },
                {
                    "disabilityActionType": "INCREASE",
                    "name": "Musculoskeletal - Foot",
                    "classificationCode": "8994",
                    "serviceRelevance": "Service in Vietnam War",
                    "approximateDate": "12-1972",
                    "ratedDisabilityId": "1234567",
                    "diagnosticCode": 9999,
                    "isRelatedToToxicExposure": false,
                    "exposureOrEventOrInjury": "Agent Orange"
                },
                {
                    "disabilityActionType": "NONE",
                    "name": "Traumatic Brain Injury",
                    "classificationCode": "9020",
                    "serviceRelevance": "Explanation of service connection",
                    "approximateDate": "11-04-2019",
                    "ratedDisabilityId": "1234567",
                    "diagnosticCode": 9999,
                    "isRelatedToToxicExposure": false,
                    "secondaryDisabilities": [
                        {
                            "name": "Cancer - Musculoskeletal - Elbow",
                            "disabilityActionType": "SECONDARY",
                            "serviceRelevance": "Injured left knee when brace on right knee failed"
                        }
                    ]
                }
            ],
            "treatments": [
                {
                    "treatedDisabilityNames": [
                        "Hearing Loss",
                        "Traumatic Brain Injury"
                    ],
                    "center": {
                        "name": "Center One",
                        "state": "GA",
                        "city": "Decatur"
                    },
                    "beginDate": "03-1985"
                },
                {
                    "treatedDisabilityNames": [
                        "Musculoskeletal - Foot"
                    ],
                    "center": {
                        "name": "Center One"
                    },
                    "beginDate": "03-1985"
                }
            ],
            "serviceInformation": {
                "servicePeriods": [
                    {
                        "serviceBranch": "Public Health Service",
                        "serviceComponent": "Active"
                    }
                ],
                "confinements": [
                    {
                        "approximateBeginDate": "06-11-2016",
                        "approximateEndDate": "09-11-2016"
                    },
                    {
                        "approximateBeginDate": "06-2019",
                        "approximateEndDate": "07-2019"
                    }
                ],
                "reservesNationalGuardService": {
                    "unitName": "National Guard Unit Name",
                    "component": "National Guard",
                    "unitPhone": {
                        "areaCode": "555",
                        "phoneNumber": "5555555"
                    },
                    "receivingInactiveDutyTrainingPay": "YES"
                }
            },
            "servicePay": {
                "favorTrainingPay": true,
                "favorMilitaryRetiredPay": false,
                "receivingMilitaryRetiredPay": "YES",
                "futureMilitaryRetiredPay": "NO",
                "militaryRetiredPay": {
                    "branchOfService": "Army",
                    "monthlyAmount": 840.75
                },
                "retiredStatus": "PERMANENT_DISABILITY_RETIRED_LIST",
                "receivedSeparationOrSeverancePay": "NO",
                "separationSeverancePay": {
                    "datePaymentReceived": "07-31-2018",
                    "branchOfService": "Naval Academy",
                    "preTaxAmountReceived": 379.25
                }
            },
            "directDeposit": {
                "accountType": "CHECKING",
                "accountNumber": "123123123123",
                "routingNumber": "123456789",
                "financialInstitutionName": "Chase",
                "noAccount": false
            }
        }
    }
}```


## What areas of the site does it impact?
	modified:   modules/claims_api/config/schemas/v2/526.json
	modified:   modules/claims_api/config/schemas/v2/request_bodies/disability_compensation/example.json
	modified:   modules/claims_api/config/schemas/v2/request_bodies/disability_compensation/request.json
	modified:   modules/claims_api/lib/claims_api/v2/disability_compensation_pdf_mapper.rb
	modified:   modules/claims_api/spec/fixtures/v2/veterans/disability_compensation/form_526_json_api.json
	modified:   modules/claims_api/spec/lib/claims_api/v2/disability_compensation_evss_mapper_spec.rb
	modified:   modules/claims_api/spec/lib/claims_api/v2/disability_compensation_pdf_mapper_spec.rb
	modified:   spec/support/schemas/claims_api/v2/forms/disability/submission.json

## Acceptance criteria

- [X]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [X]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

### CODE REVIEW

File: modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb [hunk 1]
```diff
@@ -20,14 +20,13 @@ class DisabilityCompensationController < ClaimsApi::V2::Veterans::Base
...
 
         def submit # rubocop:disable Metrics/MethodLength
-          auto_claim = ClaimsApi::AutoEstablishedClaim.create(
+          auto_claim = ClaimsApi::AutoEstablishedClaim.first_or_create(
             status: ClaimsApi::AutoEstablishedClaim::PENDING,
             auth_headers:,
             form_data: form_attributes,
             cid: token.payload['cid'],
             veteran_icn: target_veteran.mpi.icn
           )
```
Comment (stiehlrod) [inline modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb:29]: Neither the create method or [this search for an existing claim work working on several of my test cases](https://github.com/department-of-veterans-affairs/vets-api/blob/API-29607-handle-null-values/modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb#L33), so I wanted to discuss the idea or find or create. @FonzMP

File: modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb [hunk 2]
```diff
@@ -138,10 +136,6 @@ def get_pdf_data
             data: {}
           }
         end
-
-        def benefits_doc_api
-          ClaimsApi::BD.new
-        end
       end
```
Comment (stiehlrod) [inline modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb:139]: not being used

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb [hunk 3]
```diff
@@ -372,7 +358,7 @@ def validate_form_526_military_retired_pay!
         receiving_attr = form_attributes.dig('servicePay', 'receivingMilitaryRetiredPay')
         future_attr = form_attributes.dig('servicePay', 'futureMilitaryRetiredPay')
 
-        return if receiving_attr.nil? || future_attr.nil?
+        return if receiving_attr.nil?
         return unless receiving_attr == future_attr
```
Comment (stiehlrod) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:362]: future pay is nullable

File: modules/claims_api/lib/claims_api/v2/disability_compensation_evss_mapper.rb [hunk 4]
```diff
@@ -88,16 +78,6 @@ def veteran_meta
...
-      # Convert 12-05-1984 to 1984-12-05 for Docker container
-      def format_service_periods(service_period_dates)
-        service_period_dates.each do |sp_date|
-          begin_year = Date.strptime(sp_date[:activeDutyBeginDate], '%m-%d-%Y')
-          end_year = Date.strptime(sp_date[:activeDutyEndDate], '%m-%d-%Y')
-          sp_date[:activeDutyBeginDate] = begin_year.strftime('%Y-%m-%d')
-          sp_date[:activeDutyEndDate] = end_year.strftime('%Y-%m-%d')
-        end
-      end
     end
```
Comment (stiehlrod) [inline modules/claims_api/lib/claims_api/v2/disability_compensation_evss_mapper.rb:81]: Removed per this convo: https://oddball.slack.com/archives/C019XKGS1CN/p1695844255845029

File: modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb (hunk 1 above)
Comment (FonzMP) [reply modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb:29]: can you elaborate a bit? I was receiving the expected behavior from this section. Curious as to what you were experiencing. what were you getting from the create that made you want to switch to first_or_create? and above you say find or create but here you have first or create?

File: modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb (hunk 2 above)
Comment (FonzMP) [reply modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb:139]: we have access to this in the application controller in v2 anywho. same method name and all, so even if we were to use it, this is just duplication.

File: modules/claims_api/spec/requests/v2/veterans/disability_compensation_request_spec.rb [hunk 5]
```diff
@@ -1805,6 +1876,18 @@
...
+            mock_ccg(scopes) do |auth_header|
+              json = JSON.parse(data)
+              json['data']['attributes']['serviceInformation']['servicePeriods'][0]['activeDutyBeginDate'] = nil
+              data = json.to_json
+              post submit_path, params: data, headers: auth_header
+              expect(response).to have_http_status(:ok)
+            end
+          end
+        end
+
```
Comment (stiehlrod) [inline modules/claims_api/spec/requests/v2/veterans/disability_compensation_request_spec.rb:1890]: @FonzMP Would you mind pulling this down and running this test?

File: modules/claims_api/spec/requests/v2/veterans/disability_compensation_request_spec.rb [hunk 6]
```diff
@@ -1805,6 +1876,18 @@
           end
         end
 
+        context 'when the activeDutyBeginDate are null' do
+          it 'responds with a 422', skip: 'needs 2nd opinion' do
+            mock_ccg(scopes) do |auth_header|
```
Comment (stiehlrod) [inline modules/claims_api/spec/requests/v2/veterans/disability_compensation_request_spec.rb:1881]: @FonzMP Would you mind pulling this down and running this test? Take a look at the response.body, do you get a PG error regarding a null claim_id?

Comment (mchristiansonVA) [reply modules/claims_api/spec/requests/v2/veterans/disability_compensation_request_spec.rb:1881]: I am not FonzMP, but I did run this test locally and got a PG::NotNullViolation:

{\"errors\":[{\"title\":\"Internal server error\",\"detail\":\"Internal server error\",\"code\":\"500\",\"status\":\"500\",\"meta\":{\"exception\":\"PG::NotNullViolation: ERROR:  null value in column \\\"claim_id\\\" violates not-null constraint\\n

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb [hunk 7]
```diff
@@ -311,6 +310,12 @@ def validate_form_526_veteran_homelessness!
                     "defined, then 'homeless.pointOfContact' is required"
           )
         end
+
+        if international_phone_too_long?
+          raise ::Common::Exceptions::UnprocessableEntity.new(
+            detail: 'If international telephone number must be shorter than 25 characters'
```
Comment (mchristiansonVA) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:316]: Should we remove the extraneous "If" at the beginning of this detail message?

**Review by mchristiansonVA — APPROVED** [review]
One minor grammar note on a validation detail message, but otherwise tested good & swagger looks good locally as well.

**Review by mchristiansonVA — APPROVED** [review]
Pulled latest & re-tested/viewed swagger docs. LGTM!

**Pull request merged**